File: 1016-subarray-sums-divisible-by-k/1016-subarray-sums-divisible-by-k.java
class Solution {
    public int subarraysDivByK(int[] nums, int k) {

        int cnt = 0;
        int pre_sum = 0;

        Map<Integer, Integer> freq = new HashMap<>(k);
        freq.put(0, 1);

        for (int x : nums) {
            pre_sum += x;

            int m = ((pre_sum % k) + k) % k; // Ensures the remainder is non-negative

            cnt += freq.getOrDefault(m, 0);

            freq.put(m, freq.getOrDefault(m, 0) + 1);
        }

        return cnt;
    }
}

File: 1016-subarray-sums-divisible-by-k/1016-subarray-sums-divisible-by-k.py
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        cnt=0
        pre_sum=list(itertools.accumulate(nums))

        freq=defaultdict(int)
        freq[0]=1

        for x in pre_sum:
            m=(x%k)%k
            freq[m] +=1

            cnt+=freq[m]-1
        return cnt


        
       

        